### ISSUE TITLE
Anyway to load the pretrained GPT-2 model weight to levenshtein transformer?

### ISSUE THREAD

**Issue by wentinghome** [pytorch/fairseq#2435]
Hi,

I'm trying to use a small dataset (500 training examples) on levenshtein transformer to do text generation task. However, the loss maintains around 11. I think the dataset is too small, so I'm wondering if there is a way to load the pretrained GPT-2 weights to levenshtein transformer ?  Thanks.

**Comment by zodiacR** [comment]
I found a way to load pre-trained BERT models from huggingface API to fairseq transformers by substituting the name of parameters. However, since gpt2 has a different architecture from the standard transformer, my method doesn't work in this scenario. I suggest you use BERT models, or you have to modify the arch of Levenshtein transformer such that you can load gpt-2 for your tasks.

**Comment by ttzHome** [comment]
> I found a way to load pre-trained BERT models from huggingface API to fairseq transformers by substituting the name of parameters. However, since gpt2 has a different architecture from the standard transformer, my method doesn't work in this scenario. I suggest you use BERT models, or you have to modify the arch of Levenshtein transformer such that you can load gpt-2 for your tasks.

excuse me, may I ask how to load the pretrained model in huggingface  to fairseq? I want to use the bert-base-chinese in huggingface and trained it with fairseq?

**Comment by zodiacR** [comment]
> > I found a way to load pre-trained BERT models from huggingface API to fairseq transformers by substituting the name of parameters. However, since gpt2 has a different architecture from the standard transformer, my method doesn't work in this scenario. I suggest you use BERT models, or you have to modify the arch of Levenshtein transformer such that you can load gpt-2 for your tasks.
> 
> excuse me, may I ask how to load the pretrained model in huggingface to fairseq? I want to use the bert-base-chinese in huggingface and trained it with fairseq?

Hi @ttzHome, here is my solution:

```
def __init__(self, args, encoder, decoder):
        super().__init__(encoder, decoder)
        self.args = args
        self.supports_align_args = True

       pretrained_model = BertModel.from_pretrained(pretrained_model_path)
       
       # encoder
       loaded_state_dict = upgrade_state_dict_with_pretrained_weights(
                state_dict=encoder.state_dict(),
                pretrained_model=pretrained_model,
            )
       encoder.load_state_dict(loaded_state_dict, strict=True)

      # decoder
      decoder_state_dict = decoder.state_dict()
      for k in decoder_state_dict.keys():
          if k not in loaded_state_dict:
              loaded_state_dict[k] = decoder_state_dict[k]

       decoder.load_state_dict(loaded_state_dict, strict=True)

def upgrade_state_dict_with_pretrained_weights(
    state_dict: Dict[str, Any], pretrained_model: BertModel, 
) -> Dict[str, Any]:


    pretrained_state_dict = pretrained_model.state_dict()

    embed_ln = re.compile("LayerNorm\.(weight|bias)")
    self_attn = re.compile(r"layer\.(\d+)\.+attention.+((q)uery|(k)ey|(v)alue|(out)put\.dense|(output)\.LayerNorm)\.(weight|bias)")
    ffns = re.compile(r"layer\.(\d+)\.(intermediate|output)\.dense\.(weight|bias)")
    ffns_ln_p = re.compile(r"layer\.(\d+)\.output\.LayerNorm\.(weight|bias)")
        

    for key in pretrained_state_dict.keys():
        # print(key)
        if "embeddings" in key:
            if "word" in key:
                new_key = "embed_tokens.weight"
            elif "position" in key:
                new_key = 'embed_positions.weight'
            elif "type" in key:
                new_key = None
            else:
                fs_embed_ln = "layernorm_embedding.{}"
                new_key = fs_embed_ln.format(embed_ln.search(key).group(1))
        elif "attention" in key:
            # print(k)
            groups = self_attn.search(key).groups()
            fs_self_attn = "layers.{}.self_attn.{}_proj.{}"
            fs_self_attn_ln = "layers.{}.self_attn_layer_norm.{}"
            if "query" in key:
                # print(fs_self_attn.format(groups[0], groups[2], groups[-1]))
                new_key = fs_self_attn.format(groups[0], groups[2], groups[-1])
            elif "key" in key:
                # print(fs_self_attn.format(groups[0], groups[3], groups[-1]))
                new_key = fs_self_attn.format(groups[0], groups[3], groups[-1])
            elif "value" in key:
                # print(fs_self_attn.format(groups[0], groups[4], groups[-1]))
                new_key = fs_self_attn.format(groups[0], groups[4], groups[-1])
            elif "output.dense" in key:
                # print(fs_self_attn.format(groups[0], groups[5], groups[-1]))
                new_key = fs_self_attn.format(groups[0], groups[5], groups[-1])
            else:
                # print(fs_self_attn_ln.format(groups[0], groups[-1]))
                new_key = fs_self_attn_ln.format(groups[0], groups[-1])
        elif "dense" in key and "pooler" not in key:
            groups = ffns.search(key).groups()
            # print(groups)
            ffns_f1 = "layers.{}.fc1.{}"
            ffns_f2 = "layers.{}.fc2.{}"
            if "intermediate" in key:
                # print(ffns_f1.format(groups[0], groups[-1]))
                new_key = ffns_f1.format(groups[0], groups[-1])
            else:
                # print(ffns_f2.format(groups[0], groups[-1]))
                new_key = ffns_f2.format(groups[0], groups[-1])
        elif "LayerNorm" in key:
            # print(key)
            groups = ffns_ln_p.search(key).groups()
            ffns_ln = "layers.{}.final_layer_norm.{}"
            # print(groups)
            # print(ffns_ln.format(groups[0], groups[-1]))
            new_key = ffns_ln.format(groups[0], groups[-1])

        else:
            new_key = None
        
        if new_key is not None:
            # print(model[key].shape)
            # print(new_key, key)
            # print(state_dict[new_key].shape, pretrained_state_dict[key].shape)
            assert new_key in state_dict, (
                "{} Transformer encoder / decoder "
                "state_dict does not contain {}. Cannot "
                "load {} from pretrained XLM checkpoint "
                "{} into Transformer.".format(
                    str(state_dict.keys()),
                    new_key, key, pretrained_state_dict)
                )

            if new_key == "embed_tokens.weight":
                size = pretrained_state_dict[key].size(0)
                state_dict[new_key][:size] = pretrained_state_dict[key]
            else:
                assert state_dict[new_key].shape == pretrained_state_dict[key].shape
                state_dict[new_key] = pretrained_state_dict[key]

    return state_dict
```

**Comment by ttzHome** [comment]
> > > I found a way to load pre-trained BERT models from huggingface API to fairseq transformers by substituting the name of parameters. However, since gpt2 has a different architecture from the standard transformer, my method doesn't work in this scenario. I suggest you use BERT models, or you have to modify the arch of Levenshtein transformer such that you can load gpt-2 for your tasks.
> > 
> > 
> > excuse me, may I ask how to load the pretrained model in huggingface to fairseq? I want to use the bert-base-chinese in huggingface and trained it with fairseq?
> 
> Hi @ttzHome, here is my solution:
> 
> ```
> def __init__(self, args, encoder, decoder):
>         super().__init__(encoder, decoder)
>         self.args = args
>         self.supports_align_args = True
> 
>        pretrained_model = BertModel.from_pretrained(pretrained_model_path)
>        
>        # encoder
>        loaded_state_dict = upgrade_state_dict_with_pretrained_weights(
>                 state_dict=encoder.state_dict(),
>                 pretrained_model=pretrained_model,
>             )
>        encoder.load_state_dict(loaded_state_dict, strict=True)
> 
>       # decoder
>       decoder_state_dict = decoder.state_dict()
>       for k in decoder_state_dict.keys():
>           if k not in loaded_state_dict:
>               loaded_state_dict[k] = decoder_state_dict[k]
> 
>        decoder.load_state_dict(loaded_state_dict, strict=True)
> 
> def upgrade_state_dict_with_pretrained_weights(
>     state_dict: Dict[str, Any], pretrained_model: BertModel, 
> ) -> Dict[str, Any]:
> 
> 
>     pretrained_state_dict = pretrained_model.state_dict()
> 
>     embed_ln = re.compile("LayerNorm\.(weight|bias)")
>     self_attn = re.compile(r"layer\.(\d+)\.+attention.+((q)uery|(k)ey|(v)alue|(out)put\.dense|(output)\.LayerNorm)\.(weight|bias)")
>     ffns = re.compile(r"layer\.(\d+)\.(intermediate|output)\.dense\.(weight|bias)")
>     ffns_ln_p = re.compile(r"layer\.(\d+)\.output\.LayerNorm\.(weight|bias)")
>         
> 
>     for key in pretrained_state_dict.keys():
>         # print(key)
>         if "embeddings" in key:
>             if "word" in key:
>                 new_key = "embed_tokens.weight"
>             elif "position" in key:
>                 new_key = 'embed_positions.weight'
>             elif "type" in key:
>                 new_key = None
>             else:
>                 fs_embed_ln = "layernorm_embedding.{}"
>                 new_key = fs_embed_ln.format(embed_ln.search(key).group(1))
>         elif "attention" in key:
>             # print(k)
>             groups = self_attn.search(key).groups()
>             fs_self_attn = "layers.{}.self_attn.{}_proj.{}"
>             fs_self_attn_ln = "layers.{}.self_attn_layer_norm.{}"
>             if "query" in key:
>                 # print(fs_self_attn.format(groups[0], groups[2], groups[-1]))
>                 new_key = fs_self_attn.format(groups[0], groups[2], groups[-1])
>             elif "key" in key:
>                 # print(fs_self_attn.format(groups[0], groups[3], groups[-1]))
>                 new_key = fs_self_attn.format(groups[0], groups[3], groups[-1])
>             elif "value" in key:
>                 # print(fs_self_attn.format(groups[0], groups[4], groups[-1]))
>                 new_key = fs_self_attn.format(groups[0], groups[4], groups[-1])
>             elif "output.dense" in key:
>                 # print(fs_self_attn.format(groups[0], groups[5], groups[-1]))
>                 new_key = fs_self_attn.format(groups[0], groups[5], groups[-1])
>             else:
>                 # print(fs_self_attn_ln.format(groups[0], groups[-1]))
>                 new_key = fs_self_attn_ln.format(groups[0], groups[-1])
>         elif "dense" in key and "pooler" not in key:
>             groups = ffns.search(key).groups()
>             # print(groups)
>             ffns_f1 = "layers.{}.fc1.{}"
>             ffns_f2 = "layers.{}.fc2.{}"
>             if "intermediate" in key:
>                 # print(ffns_f1.format(groups[0], groups[-1]))
>                 new_key = ffns_f1.format(groups[0], groups[-1])
>             else:
>                 # print(ffns_f2.format(groups[0], groups[-1]))
>                 new_key = ffns_f2.format(groups[0], groups[-1])
>         elif "LayerNorm" in key:
>             # print(key)
>             groups = ffns_ln_p.search(key).groups()
>             ffns_ln = "layers.{}.final_layer_norm.{}"
>             # print(groups)
>             # print(ffns_ln.format(groups[0], groups[-1]))
>             new_key = ffns_ln.format(groups[0], groups[-1])
> 
>         else:
>             new_key = None
>         
>         if new_key is not None:
>             # print(model[key].shape)
>             # print(new_key, key)
>             # print(state_dict[new_key].shape, pretrained_state_dict[key].shape)
>             assert new_key in state_dict, (
>                 "{} Transformer encoder / decoder "
>                 "state_dict does not contain {}. Cannot "
>                 "load {} from pretrained XLM checkpoint "
>                 "{} into Transformer.".format(
>                     str(state_dict.keys()),
>                     new_key, key, pretrained_state_dict)
>                 )
> 
>             if new_key == "embed_tokens.weight":
>                 size = pretrained_state_dict[key].size(0)
>                 state_dict[new_key][:size] = pretrained_state_dict[key]
>             else:
>                 assert state_dict[new_key].shape == pretrained_state_dict[key].shape
>                 state_dict[new_key] = pretrained_state_dict[key]
> 
>     return state_dict
> ```

Thanks! Let me try......
